### PR TITLE
Unit Test Framework

### PR DESCRIPTION
**Pull request by RevoluPowered**
This uses doctest and doesn't use Catch2, this was because I was advised that Catch2 would severely decrease build performance.

**This includes the following:**
- a very quick test runner of a single test by running the godot executable.
- the third party library with license file (MIT) and appropriate documentation in the third party folder.
- modifications to SCSub file to make tests part of main executable as linking tests statically can be hard (need a second opinion on this approach too)

**Benefits**
- Simple testing library
- We can remove boilerplate for test cases to run test x y z.
- Most tests will be simplified into header only tests.
- Assert could be ported into gdscript and so could test cases.
- product testing with godot would become very easy.
- less code per test (because you don't need to copy paste a test harness)

Current test runner can be executed by running ./bin/godot.x11.tools.64 --test doctest

**please note**: doctest specific arguments not handled yet, but are not required to make this work #26213 

Demo with ```--test doctest``` argument

image

Demo with backward compatibility with other core tests ```--test math```
image

### CODE REVIEW

File: main/SCsub [hunk 1]
```diff
@@ -25,7 +26,5 @@ env.CommandNoCache("#main/splash_editor.gen.h", "#main/splash_editor.png", run_i
 env.Depends("#main/app_icon.gen.h", "#main/app_icon.png")
 env.CommandNoCache("#main/app_icon.gen.h", "#main/app_icon.png", run_in_subprocess(main_builders.make_app_icon))
 
-SConscript('tests/SCsub')
```
Comment (LikeLakers2) [inline main/SCsub]: With all due respect, I'm not sure why we're not using this SCons script. Was there a error of some sort that was only fixed when you did this?

File: thirdparty/doctest/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,410 @@
+# Change Log
```
Comment (LikeLakers2) [inline thirdparty/doctest/CHANGELOG.md]: We don't need the CHANGELOG for a thirdparty library in our source.

File: thirdparty/README.md [hunk 3]
```diff
@@ -1,5 +1,11 @@
 # Third party libraries
 
+## doctest
```
Comment (LikeLakers2) [inline thirdparty/README.md]: This doctest section should go between the `cvtt` and `enet` sections, as these sections are sorted alphabetically.

File: thirdparty/doctest/CHANGELOG.md (hunk 2 above)
Comment (RevoluPowered) [reply thirdparty/doctest/CHANGELOG.md]: I'll remove this

File: thirdparty/README.md (hunk 3 above)
Comment (RevoluPowered) [reply thirdparty/README.md]: Sounds good, doing that now

**Comment by RevoluPowered** [comment]
Rebased and removed requested items, also edited readme.md to use proper capitalization.

**Comment by starry-abyss** [comment]
> This uses doctest and doesn't use Catch2, this was because I was advised that Catch2 would severely decrease build performance.

I believe doctest is not very popular, it's developed by just one guy, and relies on undefined behaviour.
IMHO, if we separate test and compile targets, then we can use whatever library we like without impacting compile times (i.e. user only compiles, and CI both runs tests and compilation).

**Comment by RevoluPowered** [comment]
@starry-abyss I was advised to not separate them and I believe that unit tests should be executable from the main executable as intended by the godot developers. This is the default behavior of the engine.

Please advise how doctest is unpopular, hacky and only developed by one guy? with examples.

Furthermore: it has 31 contributors so I don't understand how 'one guy' is equal to 31 contributors.

Static analysis rates the code quality higher than godot. It isn't relying on undefined behavior and this requires further explanation.

**Comment by RevoluPowered** [comment]
After this is all working I will re-base my commits to only include the finished code.

**Comment by Anutrix** [comment]
@Gordo ☢ Your errors are because noexcept specifier. You suppress it for MSVC here https://github.com/godotengine/godot/blob/7c788945a20a315d9f5e8c955f70d2bbee09f18a/thirdparty/doctest/doctest.h#L3224 but I couldn't find it for gcc.

**Comment by RevoluPowered** [comment]
Applying '-Wno-noexcept-type' can fix this issue. I am looking for a way to not employ this across the entire code base. I will put an edit into the library temporarily - but I will generate the edit the official way for the library as I don't want to mess the library up.

I will then link the fork in the dependencies. 

Note: this library does require this, the only reason is this is a legacy version of doctest. When 4.0 is released, no modifications or warning suppression will be required. Therefore we will be able to use the normal official most up to date version of catch without this warning.

**Comment by RevoluPowered** [comment]
Ah I have found the issue. 

We aren't properly setting our available C++ version in GCC as per our standard.

the third party library is checking __cplusplus is >= 201103 which is enabling features we have specifically disabled. This could be their fault or ours.

Bizarrely without the C++11 flag applied to cmake even their own compilation fails.

```cpp
// =================================================================================================
// == FEATURE DETECTION ============================================================================
// =================================================================================================

#if __cplusplus >= 201103L
#ifndef DOCTEST_CONFIG_WITH_DELETED_FUNCTIONS
#define DOCTEST_CONFIG_WITH_DELETED_FUNCTIONS
#endif // DOCTEST_CONFIG_WITH_DELETED_FUNCTIONS
#ifndef DOCTEST_CONFIG_WITH_RVALUE_REFERENCES
#define DOCTEST_CONFIG_WITH_RVALUE_REFERENCES
#endif // DOCTEST_CONFIG_WITH_RVALUE_REFERENCES
#ifndef DOCTEST_CONFIG_WITH_NULLPTR
#define DOCTEST_CONFIG_WITH_NULLPTR
#endif // DOCTEST_CONFIG_WITH_NULLPTR
#ifndef DOCTEST_CONFIG_WITH_LONG_LONG
#define DOCTEST_CONFIG_WITH_LONG_LONG
#endif // DOCTEST_CONFIG_WITH_LONG_LONG
#ifndef DOCTEST_CONFIG_WITH_STATIC_ASSERT
#define DOCTEST_CONFIG_WITH_STATIC_ASSERT
#endif // DOCTEST_CONFIG_WITH_STATIC_ASSERT
#ifndef DOCTEST_CONFIG_WITH_VARIADIC_MACROS
#define DOCTEST_CONFIG_WITH_VARIADIC_MACROS
#endif // DOCTEST_CONFIG_WITH_VARIADIC_MACROS
#endif // __cplusplus >= 201103L
```

Simply commenting out that section fixes the issue with 9.1.0.

**Comment by akien-mga** [comment]
It's ok for libraries to use C++11. If you rebase on current master the codebase builds with C++11 support, so it should work fine.

**Comment by RevoluPowered** [comment]
@akien-mga okay, if that's now allowed I will do a rebase and restore the latest version of the library since C++11 is now enabled and usable.

File: main/tests/test_main.cpp [hunk 4]
```diff
@@ -69,8 +71,20 @@ const char **tests_get_names() {
...
+	test_context.setOption("no-breaks", true);
+	test_context.run(); // run tests
 
-	if (p_test == "string") {
+	if (test_context.shouldExit())
+		return NULL;
+
+	return NULL;
+
+	/*if (p_test == "string") {
```
Comment (jonbonazza) [inline main/tests/test_main.cpp]: I'm not sure what the project maintainers' stances are on this, but personally, I find that including commented out code into commits is... "unclean." After all, we have a version control system to see file history and revert things. I would suggest removing all commented out code from this PR.

Comment (LikeLakers2) [reply main/tests/test_main.cpp]: As far as I understand it, yes, the godot team typically doesn't want commented-out code.

That said, this PR still looks like it's a work-in-progress, so I wouldn't worry about the commented-out code just yet. When it gets to the point that this PR is considered finished and ready for review, then we can concern ourselves with clean code. :)

Comment (RevoluPowered) [reply main/tests/test_main.cpp]: Yes, this is most definitely a WIP.

Hence the commented out code. DW rebase is my friend. :)

**Comment by RevoluPowered** [comment]
Conflict because of new string method #25090 test case 35.

I will port this test soon.

**Comment by RevoluPowered** [comment]
Next patch adds compatibility for existing tests written in pure C++ rather than with doctest.

new arguments are as follows:
runs the existing tests which are not ported yet to TEST_CASE like before.
**./bin/godot.x11.tools.64 --test physics_2d**
this runs the doctest based tests:
**./bin/godot.x11.tools.64 --test doctest**

File: main/tests/test_main.cpp (hunk 4 above)
Comment (RevoluPowered) [reply main/tests/test_main.cpp]: Resolved, closing. 

Rebase will happen at end.

**Comment by RevoluPowered** [comment]
@akien-mga please can you review this?

**Comment by hpvb** [comment]
This looks good to me, it will simplify test writing a bit so maybe we will get more. And the header is tiny.

**Comment by Xrayez** [comment]
> Already planning on supporting the doctest arguments properly but this will take some time to do.

See #26213 to be implemented first perhaps, also see my little test endeavours in #32387.

**Comment by RevoluPowered** [comment]
This is best to wait until 4.0 as we will probably have to write a lot of tests then, but that requires it to be approved by core team.

**Comment by RevoluPowered** [comment]
**Later expansion for GDScript:**
After discussion with @iFire the way to handle the 'project based testing' methodology is to use a disabled doctest test, which is only enabled with arguments --test_project=/somepathtotestproject/ then the gdscript test will invoke the gdscript code with calls to the gdscript declared tests. 

This means when we implement this, you can add tests at runtime for games, without having to recompile the engine from nothing.

I believe this expansion into gdscript needs discussed and is something for godot proposals.

**Comment by follower** [comment]
FWIW, in case anyone ends up here via a search for Catch2 & Godot or GDNative...

I recently integrated Catch2 into a GDNative module for someone who was trying to "run Godot code outside Godot" but had run into issues.

The example repository is here:

 * <https://gitlab.com/RancidBacon/godoterrortest> (name relates to original issue not the Catch2 integration)

 * If you want to try adding Catch2 yourself, the specific commits for adding Catch2 support are:

    * [Add the Catch2 single header implementation](https://gitlab.com/RancidBacon/godoterrortest/-/commit/baba50b5d4ac702105ad63d6d620f5f596ca6f33)

    * [Add start of Catch2 "own main" implementation. ](https://gitlab.com/RancidBacon/godoterrortest/-/commit/0a18b7a6212b02dcbf43566f490cfea22d6511f0) (This & next commit are the key part for not having Catch2 control the test process.)

    * [Add a basic Catch2 test runner implementation](https://gitlab.com/RancidBacon/godoterrortest/-/commit/e22d7b602839931379e612d9cd38316c574b21f9)

    * [Run the tests (once added) on plugin/addon init.](https://gitlab.com/RancidBacon/godoterrortest/-/commit/4ec87bc056a04bf11ee0decce6ded674b57119c1)

    * [Add test cases: 1 to test Catch2 operation; 1 to test Godot.](https://gitlab.com/RancidBacon/godoterrortest/-/commit/45014fd535a80926ad334c55e4633d085f568a87)

**Comment by aaronfranke** [comment]
@RevoluPowered Is this still desired? If so, due to the size of this feature, you should open a [proposal](https://github.com/godotengine/godot-proposals/issues) which explains example use cases and how this approach will solve the problem.

**Comment by RevoluPowered** [comment]
> @RevoluPowered Is this still desired? If so, due to the size of this feature, you should open a [proposal](https://github.com/godotengine/godot-proposals/issues) which explains example use cases and how this approach will solve the problem.

We discussed at length with reduz and akien and it was to be merged for 4.0 (at godot con), we are looking to see if another route is practical for 4.0, doctest is still better than no test framework for core, so I was looking for a way around the shared libraries not working for test discovery.

**Comment by aaronfranke** [comment]
Yes, but if formalized as a proposal, it also allows you/reduz/akien to organize your thoughts in one place, and it gives everyone a chance to pitch in to the discussion. Having a proposal is more than just deciding whether or not it should be approved. For this topic, there are many approaches and therefore it's worth discussing openly with the community to collect ideas. Writing a proposal also forces you to think in the context of [best practices for engine contributors](https://docs.godotengine.org/en/latest/community/contributing/best_practices_for_engine_contributors.html).

**Comment by RevoluPowered** [comment]
If I get free time I will open a proposal, I'll need to abandon this PR in favor of prioritizing FBX, such a waste.

**Comment by fire** [comment]
> Can someone unabandon #30743, feel like I'm beating a dead horse? it's important it's just I need to focus on FBX. Semi-demoralised though since we discussed it at length at godotcon...

**Comment by neikeq** [comment]
What needs to be done to salvage this?

**Comment by RevoluPowered** [comment]
@neikeq in order to make this PR work we need to update doctest and rebase on 4.0

**Comment by akien-mga** [comment]
Superseded by #40148.

**Pull request closed without merging**